File: src/main/java/blockgame/world/generators/HellGenerator.java
package blockgame.world.generators;

import blockgame.block.BlockRegistry;
import blockgame.block.TilePalette;
import blockgame.block.TileState;
import org.spongepowered.noise.Noise;
import org.spongepowered.noise.NoiseQuality;
import blockgame.world.World;
import blockgame.world.generators.decorator.DungeonDecorator;
import blockgame.world.generators.decorator.IDecorator;
import blockgame.world.generators.decorator.OreDecorator;
import blockgame.world.generators.decorator.TreeDecorator;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.function.Predicate;

public class HellGenerator implements IGenerator {
        private static final List<IDecorator> _decorators = new ArrayList<>();
        private static final Predicate<TileState> REPLACE_ONLY_STONE = (TileState t) -> t.block == BlockRegistry.STONE;
        private static final Predicate<TileState> REPLACE_NOT_WATER = (TileState t) -> t.block != BlockRegistry.WATER && t.block != BlockRegistry.AIR;
        private static final Predicate<TileState> STAY_ON_DIRT = (TileState t) -> t.block == BlockRegistry.DIRT;
        private static final TileState AIR = new TileState(BlockRegistry.AIR);
        private static final TileState GRASS = new TileState(BlockRegistry.GRASS);
        private static final TileState STONE = new TileState(BlockRegistry.STONE);
        private static final TileState OBSIDIAN = new TileState(BlockRegistry.OBSIDIAN);
        private static final TileState LAVA_OBSIDIAN = new TileState(BlockRegistry.LAVA_OBSIDIAN);
        private static final TileState LAVA = new TileState(BlockRegistry.LAVA);
        private static final TileState DIRT = new TileState(BlockRegistry.DIRT);
        private static final TileState BORDERSTONE = new TileState(BlockRegistry.BORDERSTONE);

        public HellGenerator() {
            _decorators.add(new TreeDecorator(
                    new TileState(BlockRegistry.LOG),
                    new TileState(BlockRegistry.AIR), 4, STAY_ON_DIRT));
            _decorators.add(new OreDecorator(new TileState(BlockRegistry.COAL_ORE), 12, 12, 78, 20, REPLACE_ONLY_STONE ));
            _decorators.add(new OreDecorator(new TileState(BlockRegistry.IRON_ORE), 8, 4, 64, 12, REPLACE_ONLY_STONE));
            _decorators.add(new OreDecorator(new TileState(BlockRegistry.GOLD_ORE), 4, 4, 32, 8, REPLACE_ONLY_STONE));
            _decorators.add(new OreDecorator(new TileState(BlockRegistry.DIAMOND_ORE), 6, 4, 16, 8, REPLACE_ONLY_STONE));
            _decorators.add(new OreDecorator(new TileState(BlockRegistry.AIR), 6, 4, 64, 64, REPLACE_ONLY_STONE));
            _decorators.add(new DungeonDecorator(2, 16, 32));
        }

    public void generate(World world, int cX, int cZ) {
        for (int x = 0; x < 16; x++) {
            for (int z = 0; z < 16; z++) {
                for (int y = 0; y < 128; y++) {
                    int height = 32;
                    double nX = world.adjustChunk(cX, x) / 16.0;
                    double nZ = world.adjustChunk(cZ, z) / 16.0;
                    double h = 1.25 * Noise.valueCoherentNoise3D(nX / 4, 0.0, nZ / 4, world.getSeed(), NoiseQuality.FAST);
                    h += 0.25 * Noise.valueCoherentNoise3D(nX, 0.0, nZ, world.getSeed(), NoiseQuality.FAST);
                    h += 0.05 * Noise.valueCoherentNoise3D(nX*4, 0.0, nZ*4, world.getSeed(), NoiseQuality.FAST);
                    h = h-0.45;
                    h = Math.pow(h, 2.0);
                    height += (int)(h * 127.0);
                    height = Math.max(32, Math.min(height, 127));
                    boolean sandy = height < 54;
                    boolean water = height < 48;
                    TileState tile = AIR;
                    if (water) {
                        if(y >= 49 && y <= 127) {
                            tile = AIR;
                        } else if(y >= height && y <= 48) {
                            tile = LAVA;
                        } else if(y >= height-5 && y <= height-1) {
                            tile = LAVA_OBSIDIAN;
                        } else {
                            tile = STONE;
                        }
                    } else {
                        if(y >= height+1 && y <= 127) {
                            tile = AIR;
                        } else if(y >= height-5 && y <= height) {
                            if(y == 48 || y == 49) {
                                tile = LAVA_OBSIDIAN;
                            } else if(sandy)
                                tile = OBSIDIAN;
                            else
                                tile = DIRT;
                        } else {
                            tile = STONE;
                        }
                    }
                    world.setTileAtAdjusted(cX, cZ, x, y, z, TilePalette.getTileRepresentation(tile));
                }
            }
        }

        for (int x = 0;x < 16;x++) {
            for (int z = 0;z < 16;z++) {
                world.setTileAtAdjusted(cX, cZ, x, 0, z, TilePalette.getTileRepresentation(
                        BORDERSTONE
                ));
            }
        }
    }

    @Override
    public void decorate(World world, int cX, int cZ) {
        world.random = new Random(world.getSeed());
        world.random = new Random(world.getSeed()
                + (world.random.nextInt(Integer.MAX_VALUE) * cX)
                + (world.random.nextInt(Integer.MAX_VALUE) * cZ) ^ world.getSeed());
        for (IDecorator d : _decorators) {
            d.decorate(world, cX, cZ);
        }
    }
}

File: src/main/kotlin/world/generators/HellGenerator.kt
package world.generators

import block.BlockRegistry
import block.TilePalette
import block.TileState
import org.spongepowered.noise.Noise
import org.spongepowered.noise.NoiseQuality
import world.World
import world.generators.decorator.IDecorator
import world.generators.decorator.OreDecorator
import world.generators.decorator.TreeDecorator


class HellGenerator : IGenerator {

    companion object {
        var _decorators : MutableList<IDecorator> = mutableListOf()
        val REPLACE_ONLY_STONE = {t: TileState -> t.block == BlockRegistry.STONE}
        val REPLACE_NOT_WATER = {t: TileState -> t.block != BlockRegistry.WATER && t.block != BlockRegistry.AIR}
        val STAY_ON_DIRT = {t: TileState -> t.block != BlockRegistry.DIRT}
        init {
            _decorators.add(TreeDecorator(
                TileState(BlockRegistry.LOG),
                TileState(BlockRegistry.AIR), 4, STAY_ON_DIRT))
            _decorators.add(OreDecorator(TileState(BlockRegistry.COAL_ORE), 12, 12, 78, 20, REPLACE_ONLY_STONE ))
            _decorators.add(OreDecorator(TileState(BlockRegistry.IRON_ORE), 8, 4, 64, 12, REPLACE_ONLY_STONE))
            _decorators.add(OreDecorator(TileState(BlockRegistry.GOLD_ORE), 4, 4, 32, 8, REPLACE_ONLY_STONE))
            _decorators.add(OreDecorator(TileState(BlockRegistry.DIAMOND_ORE), 6, 4, 16, 8, REPLACE_ONLY_STONE))
        }
    }

    override fun generate(world: World, cX: Int, cZ: Int) {
        for (x in 0..15) {
            for (z in 0..15) {
                for (y in 0..127) {
                    var height = 32
                    var nX = ((cX * 16.0) + x) / 16.0
                    var nZ = ((cZ * 16.0) + z) / 16.0
                    var h = 1.25 * Noise.valueCoherentNoise3D(nX / 4, 0.0, nZ / 4, world.getSeed(), NoiseQuality.BEST)
                    h += 0.25 * Noise.valueCoherentNoise3D(nX, 0.0, nZ, world.getSeed(), NoiseQuality.BEST)
                    h += 0.05 * Noise.valueCoherentNoise3D(nX*4, 0.0, nZ*4, world.getSeed(), NoiseQuality.BEST)
                    h = h-0.45
                    h = Math.pow(h, 2.0)
                    height += (h * 127.0).toInt()
                    height = Math.max(32, Math.min(height, 127))
                    var sandy = height < 54
                    var water = height < 48
                    var tile = TileState(BlockRegistry.AIR)
                    if (water) {
                        tile =
                            when (y) {
                                in 49..127 -> tile
                                in height..48 -> TileState(BlockRegistry.LAVA)
                                in height-5 until height -> TileState(BlockRegistry.LAVA_OBSIDIAN)
                                else -> TileState(BlockRegistry.STONE)
                            }
                    } else {
                        tile =
                            when (y) {
                                in height + 1..127 -> tile
                                height -> if(height in 48..49) TileState((BlockRegistry.LAVA_OBSIDIAN))
                                    else if(sandy) TileState(BlockRegistry.OBSIDIAN)
                                    else TileState(BlockRegistry.DIRT)
                                in height - 5 until height -> if(y in 48..49) TileState((BlockRegistry.LAVA_OBSIDIAN))
                                    else if(sandy) TileState(BlockRegistry.OBSIDIAN)
                                    else TileState(BlockRegistry.DIRT)
                                else -> {
                                    TileState(BlockRegistry.STONE)
                                }
                            }
                    }
                    world.setTileAtAdjusted(cX, cZ, x, y, z, TilePalette.getTileRepresentation(tile))
                }
            }
        }

        for (x in 0..15) {
            for (z in 0..15) {
                world.setTileAtAdjusted(cX, cZ, x, 0, z, TilePalette.getTileRepresentation(
                    TileState(BlockRegistry.BORDERSTONE)
                ))
            }
        }
    }

    override fun decorate(world: World, cX: Int, cZ: Int) {
        for (d in _decorators) {
            d.decorate(world, cX, cZ)
        }
    }
}